cantus_text_syllabification.py

```diff
@@ -0,0 +1,206 @@
+and other alginment details. See the README for more information.
+
+Use syllabify_text to syllabify a text string. See function docstring
+or README for more details.
+"""
+
+import re
+import logging
+from typing import Union
+
+from latin_word_syllabification import syllabify_word, split_word_by_syl_bounds
+
+EXCEPTIONS_DICT = {
+    "euouae": ["e-", "u-", "o-", "u-", "a-", "e"],
+    "israelitis": ["is-", "ra-", "e-", "li-", "tis"],
+    "israel": ["is-", "ra-", "el"],
+    "michael": ["mi-", "cha-", "el"],
+}
+
+# Pre-compiled regex patterns used in this module
+# INVALID_CHAR_REGEX matches any character not valid in Cantus DB entries
+INVALID_CHAR_REGEX = re.compile(r"[^a-zA-Z#~\{\}\[\]\|\- ]")
+# Matches a string that begins with a hyphen
+STR_BEGINS_W_HYPHEN_REGEX = re.compile(r"^\-")
+# Matches a string that ends with a hyphen
+STR_ENDS_W_HYPHEN_REGEX = re.compile(r"\-$")
+# Matches pipes and missing music sectioners ("{" and "}")
+TEXT_SECTIONER_REGEX = re.compile(r"(\|)|(\{.*?\})(?!\s*?\{)")
+
+
+def _clean_text(text: str) -> str:
+    """
+    Removes invalid characters from the text string.
+
+    text [str]: string to clean
+
+    returns [str]: cleaned string
+    """
+    return INVALID_CHAR_REGEX.sub("", text)
+
+
+def _detect_invalid_characters(text: str) -> bool:
+    """
+    Detects invalid characters in the text string.
+
+    text [str]: string to check
+
+    returns [bool]: True if invalid characters are present, False otherwise
+    """
+    return bool(INVALID_CHAR_REGEX.search(text))
+
+
+def _prepare_string_for_syllabification(word_str: str) -> "tuple[str, bool, bool]":
+    """
+    Complete preparation of a string before syllabification.
+    All letters are converted to lowercase. Hyphens are removed
+    from the beginning and end of the string, and the presence
+    of these hyphens is recorded.
+
+    word_str [str]: string to prepare
+
+    returns [tuple[str,bool,bool]]: prepared string, whether a hyphen
+        was removed from the beginning of the string, whether a hyphen
+        was removed from the end of the string
+    """
+    word_str, start_hyphen = STR_BEGINS_W_HYPHEN_REGEX.subn("", word_str)
+    word_str, end_hyphen = STR_ENDS_W_HYPHEN_REGEX.subn("", word_str)
+    return word_str, bool(start_hyphen), bool(end_hyphen)
+
+
+def _split_text_sections(text: str) -> "list[str]":
+    """
+    Splits a text string into sections based on the presence of
+    curly braces "{}", square brackets "[]",
+    and pipes "|". These special characters
+    are captured in the sections.
+
+    text [str]: text to split
+
+    returns [list[str]]: list of text sections
+    """
+    text_sections = TEXT_SECTIONER_REGEX.split(text)
+    # Remove extra spaces from sections and remove empty sections/None-type sections
+    # caused by section split.
+    text_sections = [
+        section.strip() if section is not None else "" for section in text_sections
```

Comment:
You can get a `None` from a split (I was surprised by it too when it happened in testing!). 

If you are splitting by multiple strings but also capturing the strings you are splitting on (as we are here...we want to both split on "|" and capture "|" for display), you can get nones.

```python
>>> chant_text = "Dominus {Deus} | in Sion"
>>> TEXT_SECTIONER_REGEX.split(chant_text)
['Dominus ', None, '{Deus}', ' ', '|', None, ' in Sion']
```

I see what you are saying about the `filter` call so that makes sense to change. Now that I think about it, this would work in one line:

```python
>>>  text_sections = [x.strip() for x in text_sections if (x and  x.strip())]
```

Does that seem good?

Reply:
I guess the reason you're getting `None` is using too many capturing groups, so it's inserting the non-matching groups as well. (That's what [Stackoverflow tellls me anyway](https://stackoverflow.com/questions/11320231/re-split-multiple-arguments-or-returns-none-python))

How's this?

`TEXT_SECTIONER_REGEX = re.compile(r"(\||\{\w+?\}(?!\*?\{))")`

I just removed the capturing group and put the `|` matcher directly against the curly matcher within one capture group. It doesn't do anything with extra spaces, so you'll still get those, but it *might* get rid of the `None`s?

Reply:
With some additional fiddling, got a regex that yields no `None`s on the split and then can be filtered with a list comprehension. Problem was indeed too many capture groups.